src/overlay-manager.coffee

```diff
@@ -1,38 +1,101 @@
+        @renderOverlay(state, decorationId, overlay)
+        @cacheOverlayState(decorationId, overlay)
+
+    for id, {overlayNode} of @overlaysById
       unless state.content.overlays.hasOwnProperty(id)
-        delete @overlayNodesById[id]
+        delete @overlaysById[id]
         overlayNode.remove()
 
-    return
+    @cacheEditorDimensions(state)
+
+  overlayStateIsEqual: (decorationId, overlay) ->
+    return false unless @overlaysById[decorationId]?
+    @overlaysById[decorationId].itemWidth is overlay.itemWidth and
+      @overlaysById[decorationId].itemHeight is overlay.itemHeight and
+      @overlaysById[decorationId].contentMargin is overlay.contentMargin and
+      @overlaysById[decorationId].pixelPosition?.top is overlay.pixelPosition?.top and
+      @overlaysById[decorationId].pixelPosition?.left is overlay.pixelPosition?.left
+
+  cacheOverlayState: (decorationId, overlay) ->
+    return unless @overlaysById[decorationId]?
+    @overlaysById[decorationId].itemWidth = overlay.itemWidth
+    @overlaysById[decorationId].itemHeight = overlay.itemHeight
+    @overlaysById[decorationId].contentMargin = overlay.contentMargin
+    @overlaysById[decorationId].pixelPosition = overlay.pixelPosition
+
+  cacheEditorDimensions: (state) ->
+    @cachedEditorDimensions =
+      lineHeight: @presenter.lineHeight
+      contentFrameWidth: @presenter.contentFrameWidth
+      editorTop: @presenter.boundingClientRect?.top
+      editorLeft: @presenter.boundingClientRect?.left
+      editorWidth: @presenter.boundingClientRect?.width
+      windowWidth: @presenter.windowWidth
+      windowHeight: @presenter.windowHeight
+      scrollTop: state.content.scrollTop
+      scrollLeft: state.content.scrollLeft
+
+  editorDimensionsAreEqual: (state) ->
+    return false unless @cachedEditorDimensions?
+    @cachedEditorDimensions.lineHeight is @presenter.lineHeight and
+      @cachedEditorDimensions.contentFrameWidth is @presenter.contentFrameWidth and
+      @cachedEditorDimensions.editorTop is @presenter.boundingClientRect?.top and
+      @cachedEditorDimensions.editorLeft is @presenter.boundingClientRect?.left and
+      @cachedEditorDimensions.editorWidth is @presenter.boundingClientRect?.width and
+      @cachedEditorDimensions.windowWidth is @presenter.windowWidth and
+      @cachedEditorDimensions.windowHeight is @presenter.windowHeight and
+      @cachedEditorDimensions.scrollTop is state.content.scrollTop and
+      @cachedEditorDimensions.scrollLeft is state.content.scrollLeft
+
+  measureOverlays: ->
+    for decorationId, {item} of @overlaysById
+      @measureOverlay(decorationId, item)
 
-  renderOverlay: (state, decorationId, item, pixelPosition) ->
+  measureOverlay: (decorationId, item) ->
+    contentMargin = parseInt(getComputedStyle(item)['margin-left']) ? 0
+    @presenter.setOverlayDimensions(decorationId, item.offsetWidth, item.offsetHeight, contentMargin)
+
+  renderOverlay: (state, decorationId, {item, pixelPosition}) ->
```

Comment:
Most of the logic in this method should probably be moved into the presenter. You shouldn't need to cache anything in this component other than the last rendered overlay state. That way, the presenter can decide when this overlay state needs to be updated based on dimensions that get written into it and you don't need to cache a bunch of stuff at this layer.